crates/bounded_planet_core/src/unit_selection.rs

```diff
@@ -0,0 +1,207 @@
+    
+    if mouse_button_inputs.just_pressed(MouseButton::Left) {
+        // If the user clicks at all, clear the selection state. Aka, deselect whatever is selected
+        selection_state.swap_and_clear_current();
+        
+        // TODO: properly figure out _which_ camera was intended to be used
+        // Search for a camera with the default 3d camera name, which should be the "main" camera in a 3d game
+        let mut camera_query_iter = camera_query.iter();
+        let (camera, camera_transform) = camera_query_iter.into_iter()
+            .find(|&(camera, _)| {
+                // If Some(name), return name == base::camera::CAMERA3D. If None, return false
+                camera.name.as_ref().map_or(false, |name| name == base::camera::CAMERA3D)
+            })
+            .expect("Unable to find a main camera! Do you not have a Camera3dComponents in use?");
+
+        let window = windows.get(camera.window)
+            .expect("Unable to find the window of the 3d camera! Are you running headless yet trying to do screen-based unit selection?");
+        let screen_size = Vec2::new(window.width as f32, window.height as f32);
+
+        // Turn the cursor position on the screen into a position in world space (but still on the camera plane)
+        let point = math::screen_to_world(
+            state.cursor_position,
+            screen_size,
+            *camera_transform.value(),
+            camera.projection_matrix,
+        );
+
+        // Construct a direction vector starting from the camera's location in world space, passing through the cursor position in world space
+        let direction = point - camera_transform.translation();
+        // Construct the ray starting at the screen point in world space, heading in the computed direction
+        let ray = Ray::new(point.into_rapier(), direction.into_rapier());
+
+        if let Some((hit_collider, _, _)) = query_pipeline.cast_ray(&colliders, &ray, std::f32::MAX) {
+            // The raycast hit something!
+
+            // Search for an entity with the same collider handle as the collider we hit
+            let mut colliders_query_iter = colliders_query.iter();
+            let query_search = colliders_query_iter.into_iter()
+                .find(|&(_, collider_handle)| {
+                    collider_handle.handle() == hit_collider
+                });
+
+            let (collided_entity, _) = query_search
+                .expect("There was a raycast that hit an entity, but we can't find it in a query with collider handles! Somehow it got lost...");
+
+            selection_state.currently_selected = Some(collided_entity);
+            selection_events.send(SelectionEvent {
+                selected: Some(collided_entity)
+            });
+        } else {
+            // If there was no raycast collision :(
+
+            selection_state.currently_selected = None;
+            selection_events.send(SelectionEvent {
+                selected: None
+            });
```

Comment:
This kind of state should be attached to the `Entity` as a component, rather than hidden away in resources. I'd probably just emit events from this system, and have another system which consumes those events and tags the appropriate entities. This has a couple of advantages:
 - Multiple unit selection is possible.
 - `highlight_selected` system doesn't have to lookup `Entity` references.
 - Altenative selection modes (e.g. press shift to add to selection) can be handled by this other system.

This isn't really the core of what this PR is about though, so once again I'm ok with just a `todo` for this (I may even claim it myself).

Reply:
#68 is now open